### PR TITLE
Cross-platform support (tested for Windows 10)

### PR DESCRIPTION
**Pull request by marcus-wirtz-snkeos**
Trainings are running without problems on Windows 10, it should be checked however if there are issues in Linux (there shouldn't) before merging in master.

Best,
Marcus

### CODE REVIEW

**Comment by FabianIsensee** [comment]
Hi Marcus, great job! I will test it out on Linux and see if there are problems.
There is one issue, however: I cannot maintain windows compatibility looking forward. We have no Windows machine to test things on (and not the personnel to do it anyways). Therefore, once I merged this PR into master, I suggest we create a new tag and direct all windows users to this tag instead of the most recent one. Do you know how we can ensure that `pip install nnunet` will install a OS-compatible version?

**Comment by MRCWirtz** [comment]
Hi Fabian, 

I don't think it is best practice to automatically platform-dependently redirect 'pip install' since it should always refer to the latest released version. What do you think about raising a warning instead in setup.py if ("win" in sys.platform) and (nnunet.__version__ != '1.7.6') with a message like: "We don't maintain nnunet for Windows platforms except for version 1.7.6. If you run into problems, please consider to install via: \npip install nnunet==1.7.6".

To access versions via module.__version__ (anyways a good thing to have!), I would recommend to follow the reply from Zooko in:
https://stackoverflow.com/questions/458550/standard-way-to-embed-version-into-python-package

Best,
Marcus

**Comment by FabianIsensee** [comment]
Hi sorry for dragging this out so long. I will merge soon, promise haha :-D

**Comment by MRCWirtz** [comment]
Haha, no probs, better to check carefully that it doesn't destroy something important :-P

**Comment by KamalRajSinghIITR** [comment]
Hi All
Can any one suggest me that how I can use nnU-Net in Anaconda (Windows 10)?
Thank you so much in advance.

**Comment by MRCWirtz** [comment]
Hi @KamalRajSinghIITR, from your question it is a little hard to tell where your issue is. If you want to have Windows support right now, you don't have a choice rather than install it from my fork: https://github.com/MRCWirtz/nnUNet-1
Just go in your anaconda virtual environment and run 
- `git clone https://github.com/MRCWirtz/nnUNet-1.git` (or via ssh: `git clone git@github.com:MRCWirtz/nnUNet-1.git`)
- `pip install .` (within the nnunet folder of course)

For the setup you can follow the usual readme.md from this repository (regarding the environment variables you have to use Windows path structure of course) . Hope I could help.
Best, Marcus

**Comment by FabianIsensee** [comment]
@MRCWirtz is there a specific reason you replaced all the calls to join (from batchgenerators) with os.path.join? join in batchgenerators is just a shortcut to os.path.join so that we don't have to type it out all the time

File: nnunet/training/network_training/nnUNetTrainer.py [hunk 1]
```diff
@@ -296,8 +296,7 @@ def plot_network_architecture(self):
             if torch.cuda.is_available():
                 torch.cuda.empty_cache()
 
-    def save_debug_information(self):
```
Comment (FabianIsensee) [inline nnunet/training/network_training/nnUNetTrainer.py:299]: Please revert the change where you integrate `save_debug_information` into `run_training`. We need it as a separate function because we can only execute this once per node when running DDP

File: nnunet/training/network_training/nnUNetTrainerV2_DDP.py [hunk 2]
```diff
@@ -320,9 +319,6 @@ def run_training(self):
         we also need to make sure deep supervision in the network is enabled for training, thus the wrapper
         :return:
         """
-        if self.local_rank == 0:
```
Comment (FabianIsensee) [inline nnunet/training/network_training/nnUNetTrainerV2_DDP.py:323]: Please leave this in here. Otherwise each worker will save this

File: nnunet/utilities/file_endings.py [hunk 3]
```diff
@@ -17,9 +17,11 @@
 
 
 def remove_trailing_slash(filename: str):
-    while filename.endswith('/'):
-        filename = filename[:-1]
-    return filename
+    return os.path.normpath(filename)
+
+
+def get_last_folder(filename: str):
```
Comment (FabianIsensee) [inline nnunet/utilities/file_endings.py:23]: this should be foldername instead of filename and sh should check whether it actually is a folder and not a file

**Review by FabianIsensee — CHANGES_REQUESTED** [review]

**Comment by BBQtime** [comment]
Dear all, nice work! 

Do we also need to concern about the python multiprocessing problem when using windows?

[batchgenerators#windows-support-is-very-experimental](https://github.com/MIC-DKFZ/batchgenerators#windows-support-is-very-experimental)
```
if __name__ == '__main__':
    from multiprocessing import freeze_support
    freeze_support()
    main()
```

File: nnunet/run/run_training.py [hunk 4]
```diff
@@ -27,10 +27,10 @@
 
 def main():
     parser = argparse.ArgumentParser()
-    parser.add_argument("network")
-    parser.add_argument("network_trainer")
-    parser.add_argument("task", help="can be task name or task id")
-    parser.add_argument("fold", help='0, 1, ..., 5 or \'all\'')
+    parser.add_argument("-net", "--network", default='3d_fullres')
```
[the remark this answers is not in the document]
Comment (MRCWirtz) [reply nnunet/run/run_training.py:30]: Hi @FabianIsensee ,

ufff yes, it seems like there were some changes from a local working-version polluting my pull request, I am sorry for that :-(  Of course that is nonsense, I don't want to change the arguments in this way... 

In a few days I'll have some time to go over my changes one more time.

Best!

**Comment by MRCWirtz** [comment]
> @MRCWirtz is there a specific reason you replaced all the calls to join (from batchgenerators) with os.path.join? join in batchgenerators is just a shortcut to os.path.join so that we don't have to type it out all the time

Not really, it was just a rather hidden import, so I was wondering where it really came from :-D I'll revert that as well ;-)
```from batchgenerators.utilities.file_and_folder_operations import load_pickle, subfiles ```

File: nnunet/training/network_training/nnUNetTrainer.py (hunk 1 above)
Comment (MRCWirtz) [reply nnunet/training/network_training/nnUNetTrainer.py]: All changes regarding `save_debug_information()` are now reverted (I think it was a failed merge, since I not actively changed this)...

File: nnunet/training/network_training/nnUNetTrainerV2_DDP.py (hunk 2 above)
Comment (MRCWirtz) [reply nnunet/training/network_training/nnUNetTrainerV2_DDP.py]: Done.

File: nnunet/utilities/file_endings.py (hunk 3 above)
Comment (MRCWirtz) [reply nnunet/utilities/file_endings.py]: I changed the variable name and added an assertion. I didn't verify that after this changes that everything is still working, since I assume you want to test it anyways to not accidentally destroy linux support ;-)

**Comment by s-elsheikh** [comment]
@MRCWirtz 
Thanks four your effort in making this fork.

I installed nnunet-1 as instructed in a fresh conda environment (python 3.7, torch 1.8.0 + cu111)
Environment variables were set as such:

`conda env config vars set nnUNet_raw_data_base="F:\nnUNet\nnUNet_raw"`
`conda env config vars set nnUNet_preprocessed="D:\jup_nb\nnUNet\nnUNet_preprocessed"`
`conda env config vars set RESULTS_FOLDER="F:\nnUNet\nnUNet_trained_models"`

F: is HDD, D: is SDD

I wanted to try a preliminary test using the Hippocampus dataset https://github.com/MIC-DKFZ/nnUNet/blob/master/documentation/training_example_Hippocampus.md

When I run this command: nnUNet_convert_decathlon_task -i in the below mentioned variations
I get the below error.
I tried these variations:



1. `nnUNet_convert_decathlon_task -i F:/nnUNet/Task04_Hippocampus>`
2. `nnUNet_convert_decathlon_task -i F:\\nnUNet\\Task04_Hippocampus`
3. `nnUNet_convert_decathlon_task -i F:\nnUNet\Task04_Hippocampus`

I was wondering if this an error on my end. I am running Win 10. 


Error:
multiprocessing.pool.RemoteTraceback:
"""
Traceback (most recent call last):
  File "c:\users\elshikh\miniconda3\envs\nnunet\lib\multiprocessing\pool.py", line 121, in worker
    result = (True, func(*args, **kwds))
  File "c:\users\elshikh\miniconda3\envs\nnunet\lib\multiprocessing\pool.py", line 47, in starmapstar
    return list(itertools.starmap(args[0], args[1]))
  File "c:\users\elshikh\miniconda3\envs\nnunet\lib\site-packages\nnunet\experiment_planning\common_utils.py", line 27, in split_4d_nifti
    file_base = get_last_folder(filename)
  File "c:\users\elshikh\miniconda3\envs\nnunet\lib\site-packages\nnunet\utilities\file_endings.py", line 24, in get_last_folder
    assert os.path.isdir(foldername), "Provided argument 'foldername' is not a folder!"
AssertionError: Provided argument 'foldername' is not a folder!
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "c:\users\elshikh\miniconda3\envs\nnunet\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "c:\users\elshikh\miniconda3\envs\nnunet\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Users\elshikh\Miniconda3\envs\nnunet\Scripts\nnUNet_convert_decathlon_task.exe\__main__.py", line 7, in <module>
  File "c:\users\elshikh\miniconda3\envs\nnunet\lib\site-packages\nnunet\experiment_planning\nnUNet_convert_decathlon_task.py", line 54, in main
    split_4d(args.i, args.p, args.output_task_id)
  File "c:\users\elshikh\miniconda3\envs\nnunet\lib\site-packages\nnunet\experiment_planning\utils.py", line 77, in split_4d
    p.starmap(split_4d_nifti, zip(files, output_dirs))
  File "c:\users\elshikh\miniconda3\envs\nnunet\lib\multiprocessing\pool.py", line 276, in starmap
    return self._map_async(func, iterable, starmapstar, chunksize).get()
  File "c:\users\elshikh\miniconda3\envs\nnunet\lib\multiprocessing\pool.py", line 657, in get
    raise self._value
AssertionError: Provided argument 'foldername' is not a folder!

**Comment by MRCWirtz** [comment]
Hi @s-elsheikh 
sorry for replying so late to your comment, I just tried the same under Windows 10 and it just worked fine using your variation number 2 (double backslashes). Could you verify that your input folder path indeed exists by running the following in a python session?
```
import os
print(os.path.isdir('F:\\nnUNet\\Task04_Hippocampus'))
```

**Comment by s-elsheikh** [comment]
Hey @MRCWirtz 
thanks for your response. I tried checking if the directory exists, response was True.

this is the output of os.listdir:
 ['dataset.json', 'imagesTr', 'imagesTs', 'labelsTr']

I will try looking at the things you changed compared with the original repo. If you have any suggestions where this could be coming from I would be happy to hear them

**Comment by MRCWirtz** [comment]
Hi @s-elsheikh 

sorry, I originally tested the script with an empty folder that I created, however it seems like the function is not only called for the parent folder but also for all paths of the files inside... Hence I deleted the os.path.isdir() assertion and it should work just fine now. The new version is already submitted.

Apologies and happy testing!

Marcus

**Comment by s-elsheikh** [comment]
Hey @MRCWirtz 

Thanks, I am not encountering this error any more.

I appreciate your support.

File: nnunet/utilities/file_endings.py (hunk 3 above)
Comment (MRCWirtz) [reply nnunet/utilities/file_endings.py]: @FabianIsensee I had to remove the assertion again, since the code was not working anymore (see discussion with user further down in the chat): could you crosscheck if there should really be just folders as input, particularly with the script `nnUNet_convert_decathlon_task` there seems to be filepathes passed through the function.

**Comment by MRCWirtz** [comment]
Hi @FabianIsensee 
I am not sure if you are aware of the commits that I did on June 10: it should address all of your requested changes (even though I still get the message "changes requested" here in git...). If you find some time please review the new version once more.
Best
Marcus

**Comment by aghaz03** [comment]
@MRCWirtz 

I'm attempting to run nnUNet-1 (on the example Hippocampus dataset) on Windows 10. I have my path variables set in a conda environment as shown below.

(opensesame) C:\Users\aghaz\Downloads>conda env config vars list
nnUNet_raw_data_base = C:\Users\aghaz\Documents\nnUNet\nnUNet_raw_data_base
nnUNet_preprocessed = C:\Users\aghaz\Documents\nnUNet\nnUNet_preprocessed
RESULTS_FOLDER = C:\Users\aghaz\Documents\nnUNet\RESULTS_FOLDER

My data is in "C:\Users\aghaz\Documents\nnUNet\Task04_Hippocampus"

I run into an error using the nnUNet_convert_decathlon_task function.

(opensesame) C:\Users\aghaz\Documents\nnUNet>nnUNet_convert_decathlon_task -i Task04_Hippocampus

C:\Users\aghaz\Documents\nnUNet\Task04_Hippocampus

Please cite the following paper when using nnUNet:

Isensee, F., Jaeger, P.F., Kohl, S.A.A. et al. "nnU-Net: a self-configuring method for deep learning-based biomedical image segmentation." Nat Methods (2020). https://doi.org/10.1038/s41592-020-01008-z


If you have questions or suggestions, feel free to open an issue at https://github.com/MIC-DKFZ/nnUNet

Traceback (most recent call last):
  File "c:\users\aghaz\anaconda3\envs\opensesame\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "c:\users\aghaz\anaconda3\envs\opensesame\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Users\aghaz\anaconda3\envs\opensesame\Scripts\nnUNet_convert_decathlon_task.exe\__main__.py", line 7, in <module>
  File "c:\users\aghaz\anaconda3\envs\opensesame\lib\site-packages\nnunet\experiment_planning\nnUNet_convert_decathlon_task.py", line 60, in main
    split_4d(args.i, args.p, args.output_task_id)
  File "c:\users\aghaz\anaconda3\envs\opensesame\lib\site-packages\nnunet\experiment_planning\utils.py", line 65, in split_4d
    os.mkdir(curr_out_dir)
FileNotFoundError: [WinError 3] The system cannot find the path specified: 'C:\\Users\\aghaz\\Documents\\nnUNet\\nnUNet_raw_data_base\\nnUNet_raw_data\\Task004_Hippocampus\\imagesTr'FileNotFoundError: [WinError 3] The system cannot find the path specified: 'C:\\Users\\aghaz\\Documents\\nnUNet_raw_data_base\\nnUNet_raw_data\\Task004_Hippocampus\\imagesTr'

The function seems to expect a triple digit id (Task004) instead if Task04. If I change the folder names to Task004, I get the following error:

"Input folder start with TaskXX with XX being a 3-digit id: 00, 01, 02 etc"
AssertionError: Input folder start with TaskXX with XX being a 3-digit id: 00, 01, 02 etc

Any help with this error is appreciated!

Thanks,

Aryan

**Comment by MRCWirtz** [comment]
Hi @aghaz03,

yes, you are right - the further pipeline expects a 3-digit ID like "004". This issue of the 2-digit assertion in the nnUNet_convert_decathlon_task script is originating from the main branch, I guess - maybe @FabianIsensee you can comment on that?

To continue, you could try to just change the assertion of the experiment_planning/utils.py in line 45 to
> assert first_underscore == 7

(didn't test it though) and accordingly go with the 3-digit Task name.

Best
Marcus

**Comment by aghaz03** [comment]
Hi @MRCWirtz ,

I was able to get the ID working by changing "Task%03.0d_" to "Task%02.0d_" in line 55 in the utils.py file. This way the expected output folder has a two digit ID. However, now every time I run the convert function, certain data, which is in the nnUNet_raw_data folder, gets deleted and I get the following error:

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "c:\users\aghaz\anaconda3\envs\opensesame\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "c:\users\aghaz\anaconda3\envs\opensesame\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Users\aghaz\anaconda3\envs\opensesame\Scripts\nnUNet_convert_decathlon_task.exe\__main__.py", line 7, in <module>
  File "c:\users\aghaz\anaconda3\envs\opensesame\lib\site-packages\nnunet\experiment_planning\nnUNet_convert_decathlon_task.py", line 60, in main
    split_4d(args.i, args.p, args.output_task_id)
  File "c:\users\aghaz\anaconda3\envs\opensesame\lib\site-packages\nnunet\experiment_planning\utils.py", line 80, in split_4d
    p.starmap(split_4d_nifti, zip(files, output_dirs))
  File "c:\users\aghaz\anaconda3\envs\opensesame\lib\multiprocessing\pool.py", line 274, in starmap
    return self._map_async(func, iterable, starmapstar, chunksize).get()
  File "c:\users\aghaz\anaconda3\envs\opensesame\lib\multiprocessing\pool.py", line 644, in get
    raise self._value
FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\aghaz\\Documents\\nnUNet\\nnUNet_raw_data_base\\nnUNet_raw_data\\Task04_Hippocampus\\imagesTr\\Task04_Hippocampus\\imagesTr\\hippocampus_133_0000.nii.gz'

I have verified that this particular file exists before running the code. However, after I run it, all my data except for the labelsTr folder gets deleted.

Do you know if my data is in the wrong folder or is something else going on?

Thanks,

Aryan

File: nnunet/inference/predict.py [hunk 5]
```diff
@@ -22,6 +22,11 @@
 from nnunet.inference.segmentation_export import save_segmentation_nifti_from_softmax, save_segmentation_nifti
 from batchgenerators.utilities.file_and_folder_operations import *
 from multiprocessing import Process, Queue
+if 'win' in sys.platform:
```
Comment (machur) [inline nnunet/inference/predict.py:25]: "import sys" seems to be missing in this script

Comment (machur) [inline nnunet/inference/predict.py:25]: "import sys" seems to be missing in this script

**Comment by kbthornton** [comment]
Any update on when @MRCWirtz changes will be merged?
I found 76 instances of "/" and 5 instances of '/' in the current repository. If the appropriate os.path functions are used on those lines, I think that will take care of most of the problems with using nnUNet on Windows.

I recently replicated Task04 training in an Anaconda environment on a Windows 10 machine with PyTorch 1.10 and a Titan RTX GPU after changing those lines where needed.

**Comment by petsuter** [comment]
This PR uses `pathos` on Windows in [nnunet/inference/predict.py](https://github.com/MIC-DKFZ/nnUNet/pull/553/files#diff-171430529ea9e00a5c5f568af9f68b31c2dc9b0326423475e65cad53a703843d). Therefore in [setup.py](https://github.com/MIC-DKFZ/nnUNet/pull/553/files#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7) probably something like this should be added to the `install_requires` section:
```
        'pathos ; platform_system=="Windows"'
```

**Comment by FabianIsensee** [comment]
The reason this is not merged yet is because I do not feel comfortable maintaining Windows compatibility looking forward. What I could offer @MRCWirtz  is that I accept the PR in a separate branch and point potential windows users top that branch. This branch could then be regularly brought up to date with the master by someone who can actually test windows compatibility. Thoughts on that?

**Comment by kbthornton** [comment]
I understand you don't have the capability to test Windows compatibility. If the os.path functionality (replacing lines containing "/" and '/' as I mentioned above) works under Linux, I would expect it to work under Windows as well. The path issues are the only ones I've encountered so far using Windows, but I've only done training under Windows, so haven't exercised everything.

**Comment by MRCWirtz** [comment]
Hi @FabianIsensee,

sure, you can also merge in a separate branch. But why not pointing the Windows users to a specific version rather than to a branch, as we already discussed in the beginning of this thread? See:

> Hi Marcus, great job! I will test it out on Linux and see if there are problems. There is one issue, however: I cannot maintain windows compatibility looking forward. We have no Windows machine to test things on (and not the personnel to do it anyways). Therefore, once I merged this PR into master, I suggest we create a new tag and direct all windows users to this tag instead of the most recent one. Do you know how we can ensure that `pip install nnunet` will install a OS-compatible version?

Otherwise we could also just close this PR and point Windows users to my fork. 

Anyways I do not really care for which solution you decide, just go for whatever you feel most comfortable ;-)

**Comment by anabm07** [comment]
Hi @marcus-wirtz-snkeos ! thanks for the work to enable nnUNet in windows! I am currently trying to use it (only for prediction) in a computer without GPU and I run over the error "CPU not implemented". Indeed, I have seen in the function at https://github.com/marcus-wirtz-snkeos/nnUNet-Windows/blob/master/nnunet/network_architecture/neural_network.py line 118 that CPU is not supported. Do you have an explanation why? Was this difficult to implement in the windows version for some reason? 

Thanks in advance

**Comment by marcus-wirtz-snkeos** [comment]
Hi @anabm07, 
my fork of nnUNet (nnUNet-Windows) is already 139 commits behind this master. My guess is that the CPU support was enabled later and is therefore not available in nnUNet-Windows. As you can see in the discussion above, the author will not maintain Windows support looking forward and I also don't have capacities for that...

**Comment by lassoan** [comment]
FYI, prediction works perfectly on Windows, both using CPU and GPU - you only need to merge this pull request: https://github.com/MIC-DKFZ/nnUNet/pull/1155

**Comment by FabianIsensee** [comment]
I would like to point to nnunetv2 for Windows support
